src/Lookup/EntityRetrievingClosestReferencedEntityIdLookup.php

```diff
@@ -0,0 +1,236 @@
+<?php
+
+namespace Wikibase\DataModel\Services\Lookup;
+
+use Wikibase\DataModel\Entity\EntityId;
+use Wikibase\DataModel\Entity\EntityIdValue;
+use Wikibase\DataModel\Entity\PropertyId;
+use Wikibase\DataModel\Services\Entity\EntityPrefetcher;
+use Wikibase\DataModel\Snak\PropertyValueSnak;
+use Wikibase\DataModel\Snak\Snak;
+use Wikibase\DataModel\Statement\StatementListProvider;
+
+/**
+ * Service for getting the closest entity (out of a specified set),
+ * from a given starting entity. The starting entity, and the target entities
+ * are (potentially indirectly, via intermediate entities) linked by statements
+ * with a given property ID, pointing from the starting entity to one of the
+ * target entities.
+ *
+ * @since 3.10
+ *
+ * @license GPL-2.0-or-later
+ * @author Marius Hoch
+ */
+class EntityRetrievingClosestReferencedEntityIdLookup implements ReferencedEntityIdLookup {
+
+	/**
+	 * @var EntityLookup
+	 */
+	private $entityLookup;
+
+	/**
+	 * @var EntityPrefetcher
+	 */
+	private $entityPrefetcher;
+
+	/**
+	 * @var int Maximum search depth: Maximum number of intermediate entities to search through.
```

Comment:
I hope I correctly understood your comment, if not, please correct me!

Please note that this is not doing a _depth-first search_, but a _breadth-first search_, thus there's no need to make "the code stop traversing a particular branch". In this context I think it might be useful for a caller to say that they only want us to look for paths with a certain maximum length.
The entity visit limitation is just an added safety, which I don't imagine firing very often, but only in specific cases where one or more visited entities references many other entities.

I agree that the documentation for this might be improvable. I'll see what I can do there.

Reply:
I'm afraid I still don't know what this integer represents. The documentation uses the words "depth" (as in how deep a tree is) and "number" (as in the number of visited leaves) in the same sentence. For me these are two very different concepts. Which one does this integer represent?

Reply:
My understanding is that this represents the maximum *depth* of the tree. You can also think of the depth as the *number of entities above the current leaf* (i. e., intermediate entities) at any point. Does that make sense?

Reply:
So "maxDepth" is indeed the maximum depth of that particular tree. :-) Thanks!